From 2350261cc0a24addd938ec6b665768ea1db05d42 Mon Sep 17 00:00:00 2001
From: kitakar5525 <34676735+kitakar5525@users.noreply.github.com>
Date: Mon, 27 Aug 2019 19:04:56 +0900
Subject: [PATCH] mwifiex_pcie: disable parent bridge_d3

On Windows, it seems that wifi parent bridge will not enter D3 state
(stay on D0 state). This commit will disable the parent bridge D3 state.

As a side effect, it reports parent bridge is in D3hot. I'm not sure why.

```
acpi device:4b: Cannot transition to power state D0 for parent in D3hot
```

Power state of the parent bridge on SB1 before this commit:
```
$ grep -H . /sys/bus/pci/devices/0000:00:1d.3/firmware_node/{power_state,real_power_state}
/sys/bus/pci/devices/0000:00:1d.3/firmware_node/power_state:D0
/sys/bus/pci/devices/0000:00:1d.3/firmware_node/real_power_state:D3hot
```

Power state of the parent bridge on SB1 after this commit:
```
$ grep -H . /sys/bus/pci/devices/0000:00:1d.3/firmware_node/{power_state,real_power_state}
/sys/bus/pci/devices/0000:00:1d.3/firmware_node/power_state:D3hot
/sys/bus/pci/devices/0000:00:1d.3/firmware_node/real_power_state:D3hot
```

Link: https://github.com/jakeday/linux-surface/issues/420
[Wi-Fi drops after suspend · Issue #420 · jakeday/linux-surface]
---
 drivers/net/wireless/marvell/mwifiex/pcie.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/net/wireless/marvell/mwifiex/pcie.c b/drivers/net/wireless/marvell/mwifiex/pcie.c
index 6e734a83e..a94fd0e8e 100644
--- a/drivers/net/wireless/marvell/mwifiex/pcie.c
+++ b/drivers/net/wireless/marvell/mwifiex/pcie.c
@@ -233,8 +233,12 @@ static int mwifiex_pcie_probe(struct pci_dev *pdev,
 					const struct pci_device_id *ent)
 {
 	struct pcie_service_card *card;
+	struct pci_dev *parent_pdev = pci_upstream_bridge(pdev);
 	int ret;
 
+	pr_alert("DEBUG: disabling bridge_d3...\n");
+	parent_pdev->bridge_d3 = false;
+
 	pr_debug("info: vendor=0x%4.04X device=0x%4.04X rev=%d\n",
 		 pdev->vendor, pdev->device, pdev->revision);
 
-- 
2.23.0

